src/main/java/net/dries007/tfc/objects/recipes/MetalItemRecipe.java

```
@@ -0,0 +1,169 @@
+import javax.annotation.Nullable;
+
+import com.google.gson.JsonObject;
+import net.minecraft.item.ItemStack;
+import net.minecraft.item.crafting.IRecipeSerializer;
+import net.minecraft.item.crafting.IRecipeType;
+import net.minecraft.item.crafting.Ingredient;
+import net.minecraft.network.PacketBuffer;
+import net.minecraft.util.JSONUtils;
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.util.text.ITextComponent;
+import net.minecraft.util.text.TranslationTextComponent;
+import net.minecraft.world.World;
+import net.minecraftforge.common.crafting.CraftingHelper;
+
+import net.dries007.tfc.TerraFirmaCraft;
+import net.dries007.tfc.api.Metal;
+import net.dries007.tfc.objects.types.MetalManager;
+
+/**
+ * Handles mapping item stacks -> metals
+ *
+ * @see Metal
+ * @see MetalManager
+ */
+public class MetalItemRecipe implements ISimpleRecipe<ItemStackRecipeWrapper>
+{
+    public static Optional<MetalItem> getMetalItem(World world, ItemStack stack)
+    {
+        ItemStackRecipeWrapper wrapper = new ItemStackRecipeWrapper(stack);
+        return RecipeCache.INSTANCE.get(TFCRecipeTypes.METAL_ITEM, world, wrapper)
+            .map(recipe -> new MetalItem(recipe.getMetal(), recipe.getAmount()));
+    }
+
+    public static void addTooltipInfo(World world, ItemStack stack, List<ITextComponent> text)
+    {
+        getMetalItem(world, stack).ifPresent(metalItem -> {
+            text.add(new TranslationTextComponent(TerraFirmaCraft.MOD_ID + ".tooltip.metal", metalItem.getMetal().getDisplayName()));
+            text.add(new TranslationTextComponent(TerraFirmaCraft.MOD_ID + ".tooltip.units", metalItem.getAmount()));
+            text.add(metalItem.getMetal().getTier().getDisplayName());
+        });
+    }
+
+    protected final ResourceLocation id;
+    protected final Ingredient ingredient;
+    protected final ResourceLocation metal; // Have to store the resource location because at load time metals are not accessible yet
```

This should definitely be reload-time validated. There's *potential* to have a task-completion based reload listener but I'd have to look further into exactly how we should approach this.

Metals should **NOT** be silently defaulting, they should error if the metal is not specified correctly.

---

I can only validate at execution time currently. At recipe load time, metals are not yet loaded hence i can't check if it is valid or even store the `Metal` obj directly. If you find a way lmk

---

Gonna do the discussed refactor later